File: app/src/main/java/com/siavash/dualcamera/fragments/ShareFragment.java
package com.siavash.dualcamera.fragments;

import android.app.Activity;
import android.app.Fragment;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.siavash.dualcamera.Constants;
import com.siavash.dualcamera.R;
import com.siavash.dualcamera.util.StringUtil;
import com.siavash.dualcamera.util.Util;
import com.siavash.dualcamera.util.customviews.Toolbar;

import java.io.File;
import java.util.List;

import butterknife.Bind;
import butterknife.ButterKnife;

/**
 * Share or save photos to file
 * Created by siavash on 3/6/2015.
 */
public class ShareFragment extends Fragment implements Toolbar.OnBackClickListener {

    private static final String TAG = ShareFragment.class.getSimpleName();

    @Bind(R.id.toolbar) Toolbar<ShareFragment> toolbar;
    @Bind({R.id.facebook, R.id.whatsapp, R.id.telegram, R.id.instagram, R.id.line, R.id.more}) List<Button> socialNetworks;
    @Bind(R.id.share_to) TextView shareTextView;
    @Bind(R.id.photo_container) ImageView image;

    private OnFragmentChange mCallback;
    private String mImageUrl;

    private ShareFragment() {
    }

    public static ShareFragment newInstance(String imageUrl) {
        ShareFragment shareFragment = new ShareFragment();
        Bundle args = new Bundle();
        args.putString(Constants.IMAGE_URL, imageUrl);
        shareFragment.setArguments(args);
        return shareFragment;
    }

    @Override public void onAttach(Activity activity) {
        super.onAttach(activity);
        try {
            mCallback = (OnFragmentChange) getActivity();
        } catch (ClassCastException e) {
            throw new ClassCastException(activity.toString() + " must implement OnFragmentChange");
        }
    }

    @Override public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        DisplayMetrics metrics = new DisplayMetrics();
        getActivity().getWindowManager().getDefaultDisplay().getMetrics(metrics);

        int imageWidth = metrics.widthPixels;
        int imageHeight = metrics.heightPixels;

        mImageUrl = getArguments().getString(Constants.IMAGE_URL);
        Bitmap bitmap = Util.decodeSampledBitmap(Util.getFile(mImageUrl), imageWidth, imageHeight);
        image.setImageBitmap(bitmap);
    }

    @Nullable @Override public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_share, container, false);
        ButterKnife.bind(this, view);
        // Set up toolbar
        toolbar.setTitle("اشتراک گذاری");
        toolbar.setCallback(this);

        setTypefaces();
        setListeners();

        return view;
    }

    private void setTypefaces(){
        shareTextView.setTypeface(StringUtil.getFont(getActivity(), StringUtil.FONT_AFSANEH));
        for (Button button : socialNetworks){
            button.setTypeface(StringUtil.getFont(getActivity(), StringUtil.FONT_NAZANIN_BOLD));
        }
    }

    private void setListeners() {
        OnClickListener onClickListener = new OnClickListener();
        socialNetworks.get(0).setOnClickListener(onClickListener);
        socialNetworks.get(1).setOnClickListener(onClickListener);
        socialNetworks.get(2).setOnClickListener(onClickListener);
        socialNetworks.get(3).setOnClickListener(onClickListener);
        socialNetworks.get(4).setOnClickListener(onClickListener);
        socialNetworks.get(5).setOnClickListener(onClickListener);
    }

    private void shareIntent(String intentName, String socialName) {
        Intent intent = new Intent(Intent.ACTION_SEND);
        intent.setType("image/jpg");
        intent.putExtra(Intent.EXTRA_STREAM, Uri.fromFile(new File(mImageUrl)));
        generateCustomIntent(intent, intentName, socialName);
    }


    private void generateCustomIntent(Intent prototype, String appNameToShareWith, String appNameInPersian) {
        List<ResolveInfo> resInfo = getActivity().getPackageManager().queryIntentActivities(prototype, PackageManager.MATCH_DEFAULT_ONLY);
        boolean resolved = false;
        if (!resInfo.isEmpty()) {
            for (ResolveInfo resolveInfo : resInfo) {
                if (resolveInfo.activityInfo.name.contains(appNameToShareWith)) {
                    prototype.setClassName(resolveInfo.activityInfo.packageName, resolveInfo.activityInfo.name);
                    resolved = true;
                    break;
                }
            }
        }

        if (resolved) startActivity(prototype);
        else {
            Toast.makeText(getActivity(), appNameInPersian + " نصب نیست!", Toast.LENGTH_LONG).show();
        }
    }

    @Override public void goBack() {
        getActivity().onBackPressed();
    }

    private class OnClickListener implements View.OnClickListener {

        @Override
        public void onClick(View v) {
            int id = v.getId();
            if (id == socialNetworks.get(0).getId()) {
                shareIntent("facebook", "فیسبوک");
            } else if (id == socialNetworks.get(1).getId()) {
                shareIntent("whatsapp", "واتس اپ");
            } else if (id == socialNetworks.get(2).getId()) {
                shareIntent("telegram", "تلگرام");
            } else if (id == socialNetworks.get(3).getId()) {
                shareIntent("instagram", "اینستاگرام");
            } else if (id == socialNetworks.get(4).getId()) {
                shareIntent("line", "لاین");
            } else if (id == socialNetworks.get(5).getId()) {
                Intent intent = new Intent(Intent.ACTION_SEND);
                intent.setType("image/jpg");
                intent.putExtra(Intent.EXTRA_STREAM, Uri.fromFile(new File(mImageUrl)));
                startActivity(intent);
            }
        }
    }
}

File: app/src/main/kotlin/com/siavash/dualcamera/fragments/ShareFragment.kt
package com.siavash.dualcamera.fragments

import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Point
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import com.siavash.dualcamera.R
import com.siavash.dualcamera.activities.PhotoActivity
import com.siavash.dualcamera.util.*
import org.jetbrains.anko.act
import java.io.File

/**
 * Created by sia on 11/2/15.
 */
class ShareFragment : BaseFragment() {
    val socialNetworks: List<Button> by bindViews(R.id.facebook, R.id.whatsapp, R.id.telegram, R.id.instagram, R.id.line, R.id.more)
    val shareText: TextView by bindView(R.id.share_to)
    val image: ImageView by bindView(R.id.photo_container)
    lateinit var displaySize : Point

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_share, container, false)
        (act as PhotoActivity).toolbarTitle.text = "اشتراک گذاری"
        displaySize = getDisplaySize(act)
        return view
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        val bitmap = Util.decodeSampledBitmap(File(getExternalApplicationStorage(), finalImageUrl), displaySize.x, displaySize.y)
        image.setImageBitmap(bitmap)
        setTypefaces()
        setListeners()
    }

    private fun setTypefaces() {
        //fixme string utils should be modified to kotlin equivalent
        shareText.typeface = StringUtil.getFont(activity, StringUtil.FONT_AFSANEH)
        for (button in socialNetworks) {
            button.typeface = StringUtil.getFont(activity, StringUtil.FONT_NAZANIN_BOLD)
        }
    }

    private fun setListeners() {
        val onClickListener = OnClickListener()
        socialNetworks[0].setOnClickListener(onClickListener)
        socialNetworks[1].setOnClickListener(onClickListener)
        socialNetworks[2].setOnClickListener(onClickListener)
        socialNetworks[3].setOnClickListener(onClickListener)
        socialNetworks[4].setOnClickListener(onClickListener)
        socialNetworks[5].setOnClickListener(onClickListener)
    }

    private fun shareIntent(intentName: String, socialName: String) {
        val intent = Intent(Intent.ACTION_SEND)
        intent.setType("image/jpg")
        intent.putExtra(Intent.EXTRA_STREAM, Uri.fromFile(File(getExternalApplicationStorage(), finalImageUrl)))
        generateCustomIntent(intent, intentName, socialName)
    }


    private fun generateCustomIntent(prototype: Intent, appNameToShareWith: String, appNameInPersian: String) {
        val resInfo = activity.packageManager.queryIntentActivities(prototype, PackageManager.MATCH_DEFAULT_ONLY)
        var resolved = false
        if (!resInfo.isEmpty()) {
            for (resolveInfo in resInfo) {
                if (resolveInfo.activityInfo.name.contains(appNameToShareWith)) {
                    prototype.setClassName(resolveInfo.activityInfo.packageName, resolveInfo.activityInfo.name)
                    resolved = true
                    break
                }
            }
        }

        if (resolved)
            startActivity(prototype)
        else {
            Toast.makeText(activity, appNameInPersian + " نصب نیست!", Toast.LENGTH_LONG).show()
        }
    }

    private inner class OnClickListener : View.OnClickListener {
        override fun onClick(v: View) {
            val id = v.id
            if (id == socialNetworks[0].id) {
                shareIntent("facebook", "فیسبوک")
            } else if (id == socialNetworks[1].id) {
                shareIntent("whatsapp", "واتس اپ")
            } else if (id == socialNetworks[2].id) {
                shareIntent("telegram", "تلگرام")
            } else if (id == socialNetworks[3].id) {
                shareIntent("instagram", "اینستاگرام")
            } else if (id == socialNetworks[4].id) {
                shareIntent("line", "لاین")
            } else if (id == socialNetworks[5].id) {
                val intent = Intent(Intent.ACTION_SEND)
                intent.setType("image/jpg")
                intent.putExtra(Intent.EXTRA_STREAM, Uri.fromFile(File(getExternalApplicationStorage(), finalImageUrl)))
                startActivity(intent)
            }
        }
    }
}